File: js/unityfs/bundleFile.js
import {BinaryReader} from "../binaryReader";
import {lzmaDecompress} from "./utils";
import {FileType, UnityFS} from "./unityFile";
import {BinaryWriter} from "../binaryWriter";
import {lz4_decompress} from "../encoders";
import {decompressBlock} from "lz4js";

export class BundleFlags {
  static exposedAttributes = [
    'compressionType',
    'hasDirInfo',
    'blockInfoAtEnd',
    'oldWebPluginCompat',
    'blockInfoHasPadding'
  ]

  constructor(value) {
    this.compressionType = value & 0x3F;
    this.hasDirInfo = (value & 0x40) === 0x40;
    this.blockInfoAtEnd = (value & 0x80) === 0x80;
    this.oldWebPluginCompat = (value & 0x100) === 0x100;
    this.blockInfoHasPadding = (value & 0x200) === 0x200;
  }

  encode() {
    let value = this.compressionType & 0x03;
    if (this.hasDirInfo) value |= 0x40;
    if (this.blockInfoAtEnd) value |= 0x80;
    if (this.oldWebPluginCompat) value |= 0x100;
    if (this.blockInfoHasPadding) value |= 0x200;
    return value
  }
}

export class BlockFlags {
  static exposedAttributes = [
    'compressionType',
    'isStreamed'
  ];

  constructor(value) {
    this.compressionType = value & 0x3F;
    this.isStreamed = (value & 0x40) === 0x40;
  }

  encode() {
    let value = this.compressionType & 0x03;
    if (this.isStreamed) value |= 0x40;
    return value;
  }
}

export class StorageBlock {
  static exposedAttributes = [
    'compressedSize',
    'uncompressedSize',
    'flags'
  ];

  constructor(uncompressedSize, compressedSize, flags) {
    this.uncompressedSize = uncompressedSize;
    this.compressedSize = compressedSize;
    this.flags = new BlockFlags(flags);
  }
}

export class Node {
  static exposedAttributes = [
    'offset',
    'size',
    'flags',
    'path'
  ];

  constructor(offset, size, flags, path) {
    this.offset = offset;
    this.size = size;
    this.flags = flags;
    this.path = path;
  }
}

export class NodeFile {  // special support
  constructor(node, data) {
    this.node = node;
    this.data = data;
  }
}

export const CompressionType = {
  None: 0,
  LZMA: 1,
  LZ4: 2,
  LZ4HC: 3,
  LZHAM: 4
}

export class BundleFile {
  static exposedAttributes = [
    'magic',
    'version',
    'unityVersion',
    'unityRevision',
    // These *should* be present once web/raw support is introduced
    'size',
    'compressedBlockInfoSize',
    'uncompressedBlockInfoSize',
    'flags',
    'blockInfo',
    'nodes',
    'files'
  ]

  constructor(reader) {
    this.reader = reader;
  }

  parse() {
    this.magic = this.reader.readCString();
    this.version = this.reader.readUInt32();
    this.unityVersion = this.reader.readCString();
    this.unityRevision = this.reader.readCString();

    switch (this.magic) {
      case 'UnityFS':
        this.parseUnityFS();
    }
  }

  serialize() {
    if (typeof this.size == 'undefined') {
      console.error('Size required to serialize bundle');
    }
    let writer = new BinaryWriter(Number(this.size));
    writer.writeCString(this.magic);
    writer.writeUInt32(this.version);
    writer.writeCString(this.unityVersion);
    writer.writeCString(this.unityRevision);

    switch (this.magic) {
      case 'UnityFS':
        writer.write(this.serializeUnityFS());
    }

    return writer.getData();
  }

  parseUnityFS() {
    this.size = this.reader.readUInt64();
    this.compressedBlockInfoSize = this.reader.readUInt32();
    this.uncompressedBlockInfoSize = this.reader.readUInt32();
    this.flags = new BundleFlags(this.reader.readUInt32());
    if (this.magic !== 'UnityFS') {
      this.reader.read(1);  // extra byte in raw/web 6+
    }

    if (this.version >= 7) {
      this.reader.align(16);
    }
    let blockInfoData;
    if (this.flags.blockInfoAtEnd) {
      const origPos = this.reader.tell();
      this.reader.seek(this.reader.data.length - this.compressedBlockInfoSize);
      blockInfoData = this.reader.read(this.compressedBlockInfoSize);
      this.reader.seek(origPos);
    } else {
      blockInfoData = this.reader.read(this.compressedBlockInfoSize);
    }

    const onDecompress = data => {
      let reader = new BinaryReader(data);
      let dataHash = reader.read(16);
      let blockInfoCount = reader.readUInt32();
      this.blockInfo = [];
      for (let i = 0; i < blockInfoCount; i++) {
        this.blockInfo.push(new StorageBlock(
          reader.readUInt32(),
          reader.readUInt32(),
          reader.readUInt16()
        ));
      }
      this.nodes = [];
      let nodeCount = reader.readUInt32();
      for (let i = 0; i < nodeCount; i++) {
        this.nodes.push(new Node(
          reader.readUInt64(),
          reader.readUInt64(),
          reader.readUInt32(),
          reader.readCString()
        ));
      }

      let totalSize = 0;
      let uncompressedBlocks = [];
      for (let block of this.blockInfo) {
        const blockData = this.reader.read(block.compressedSize);
        let decompressedBlock = null;

        switch (block.flags.compressionType) {
          case CompressionType.None:
            decompressedBlock = blockData;
            break;
          case CompressionType.LZMA:
            decompressedBlock = lzmaDecompress(blockData, block.uncompressedSize);
            break;
          case CompressionType.LZ4:
          case CompressionType.LZ4HC:
            try {
              decompressedBlock = lz4_decompress(blockData, block.uncompressedSize);
            } catch (e) {
              console.error('error in Wasm LZ4 decompress:');
              console.error(e);
              console.error('Falling back to slower pure-JS decompressor');

              const dst = new Uint8Array(block.uncompressedSize);
              decompressBlock(blockData, dst, 0, blockData.length, 0);
              decompressedBlock = dst;
            }
            break;
          default:
            throw new Error('Unsupported compression type');
        }

        uncompressedBlocks.push(decompressedBlock);
        totalSize += block.uncompressedSize;
      }

      this.blockData = new Uint8Array(totalSize);
      let offset = 0;
      for (let block of uncompressedBlocks) {
        this.blockData.set(block, offset);
        offset += block.length;
      }
      // Extract files from blocks
      this.files = [];
      for (let node of this.nodes) {
        this.files.push(new NodeFile(
          node,
          this.blockData.slice(Number(node.offset), Number(node.offset + node.size))
        ));
      }
      this.blockData = null;

      for (let file of this.files) {
        let reader = new UnityFS(file.data);
        reader.parseHeader();
        file.type = reader.fileType;
      }
    };

    switch (this.flags.compressionType) {
      case CompressionType.None:
        onDecompress(blockInfoData);
        break;
      case CompressionType.LZMA:
        onDecompress(lzmaDecompress(blockInfoData, this.uncompressedBlockInfoSize));
        break;
      case CompressionType.LZ4:
      case CompressionType.LZ4HC:
        onDecompress(lz4_decompress(blockInfoData, this.uncompressedBlockInfoSize));
        break;
      default:
        throw new Error('Unsupported compression type');
    }

    // console.log(this.serialize());
  }

  serializeUnityFS() {
    // TODO: finish this
    // bundles shall not exceed 9 petabytes (but, knowing game developers, will)
    const writer = new BinaryWriter(Number(this.size));
    writer.writeUInt64(this.size);
    writer.writeUInt32(this.compressedBlockInfoSize);
    writer.writeUInt32(this.uncompressedBlockInfoSize);
    let copyFlags = new BundleFlags(this.flags.encode());
    copyFlags.compressionType = CompressionType.None;  // TODO: implement other compression types
    writer.writeUInt32(copyFlags.encode());

    if (this.version >= 7) {
      writer.align(16);
    }

    // Serialize files
    const newNodes = [];
    const blockData = new BinaryWriter(0);
    for (let file of this.files) {
      let data = new Uint8Array(0);
      switch (file.type) {
        case FileType.Resource:
          data = file.data;
          break;
        case FileType.Bundle:
          data = new BundleFile(new BinaryReader(file)).serialize();
          break;
        case FileType.Assets:
          data = file.serialize();
          break;
      }
      newNodes.push({
        offset: blockData.offset,
        size: data.length,
        flags: file.node.flags,
        path: file.node.path
      });
      blockData.write(data);
    }

    // Compress blocks
    // compressBlock()

    if (!this.flags.blockInfoAtEnd) {
      // Hash - 16 bytes, isn't checked so we don't bother regenerating it (instead, we just zero it)
      writer.writeUInt64(BigInt(0));
      writer.writeUInt64(BigInt(0));



      writer.writeUInt32(this.blockInfo.length);
      for (let info of this.blockInfo) {
        writer.writeUInt32(info.uncompressedSize);
        writer.writeUInt32(info.compressedSize);
        let copyBlkFlags = new BundleFlags(this.flags.encode());
        copyBlkFlags.compressionType = CompressionType.LZ4;  // TODO: ^^
        writer.writeUInt16(copyBlkFlags.encode());
      }

      writer.writeUInt32(this.nodes.length);
      for (let node of this.nodes) {
        writer.writeUInt64(node.offset);
        writer.writeUInt64(node.size);
        writer.writeUInt32(node.flags);
        writer.writeCString(node.path);
      }
    }

    return writer.getData();
  }

  parseUnityWeb() {
    if (this.version >= 6) {
      this.parseUnityFS();
      return;
    }
  }
}

File: src/unity/bundle_file.rs
use bytes::{Bytes, Buf, BytesMut, BufMut};
use crate::unity::compression::decompress;
use crate::utils::{BufExt, lz4_decompress};

#[derive(Debug)]
pub struct BundleFileHeader {
    pub magic: String,
    pub version: i32,
    pub unity_version: String,
    pub unity_revision: String,
    pub size: u64,
    pub compressed_block_info_size: u32,
    pub uncompressed_block_info_size: u32,
    pub flags: BundleFlags,
}

#[derive(Debug)]
pub struct BundleFlags {
    pub compression_type: u8,
    pub has_dir_info: bool,
    pub block_info_at_end: bool,
    pub old_web_plugin_compat: bool,
    pub block_info_has_padding: bool
}

impl BundleFlags {
    pub fn from_bits(i: u32) -> BundleFlags {
        BundleFlags {
            compression_type: (&i & 0x3f) as u8,
            has_dir_info: (&i & 0x40) == 0x40,
            block_info_at_end: (&i & 0x80) == 0x80,
            old_web_plugin_compat: (&i & 0x100) == 0x100,
            block_info_has_padding: (&i & 0x200) == 0x200
        }
    }
}

#[derive(Debug)]
pub struct BlockFlags {
    pub compression_type: u8,
    pub is_streamed: bool
}

impl BlockFlags {
    pub fn from_bits(i: u16) -> BlockFlags {
        BlockFlags {
            compression_type: (&i & 0x3f) as u8,
            is_streamed: (&i & 0x40) == 0x40
        }
    }
}

#[derive(Debug)]
pub struct BlockInfo {
    pub compressed_size: usize,
    pub uncompressed_size: usize,
    pub flags: BlockFlags
}

impl BlockInfo {

}

#[derive(Debug)]
pub struct NodeFlags {
    // TODO
}

impl NodeFlags {
    pub fn from_bits(i: u32) -> NodeFlags {
        NodeFlags {
            // TODO
        }
    }
}

#[derive(Debug)]
pub struct Node {
    pub offset: usize,
    pub size: usize,
    pub flags: NodeFlags,
    pub path: String
}

#[derive(Debug)]
pub struct StorageInfo {
    pub blocks: Vec<BlockInfo>,
    pub nodes: Vec<Node>,
    pub data_hash: Bytes,
}

#[derive(Debug)]
pub struct BundleFile {
    pub header: BundleFileHeader,
    pub storage: StorageInfo
}

impl BundleFile {
    pub fn new(data: &mut Bytes) -> BundleFile {
        BundleFile::from_bytes(data)
    }

    fn from_bytes(data: &mut Bytes) -> BundleFile {
        let header = Self::parse_header(data);
        let storage = Self::parse_storage_info(data, header.compressed_block_info_size,
                                                header.uncompressed_block_info_size, &header.flags);
        Self { header, storage }
    }

    fn parse_header(data: &mut Bytes) -> BundleFileHeader {
        let h = BundleFileHeader {
            magic: data.get_cstring(),
            version: data.get_i32(),
            unity_version: data.get_cstring(),
            unity_revision: data.get_cstring(),
            size: data.get_u64(),
            compressed_block_info_size: data.get_u32(),
            uncompressed_block_info_size: data.get_u32(),
            flags: BundleFlags::from_bits(data.get_u32())
        };
        if h.version >= 7 {
            data.align(h.size as usize, 16);
        }
        h
    }

    fn parse_storage_info(data: &mut Bytes, compressed_block_info_size: u32, uncompressed_block_info_size: u32, flags: &BundleFlags) -> StorageInfo {
        let mut decomp = decompress(
            data,
            flags.compression_type,
            compressed_block_info_size as usize,
            uncompressed_block_info_size as usize
        );
        let data_hash = decomp.slice(0..16);
        decomp.advance(16);
        let block_info_count = decomp.get_u32();
        let mut blocks = Vec::<BlockInfo>::new();
        for _ in 0..block_info_count {
            blocks.push(BlockInfo {
                compressed_size: decomp.get_u32() as usize,
                uncompressed_size: decomp.get_u32() as usize,
                flags: BlockFlags::from_bits(decomp.get_u16())
            });
        }
        let node_count = decomp.get_u32();
        let mut nodes = Vec::<Node>::new();
        for _ in 0..node_count {
            nodes.push(Node {
                offset: decomp.get_u64() as usize,
                size: decomp.get_u64() as usize,
                flags: NodeFlags::from_bits(decomp.get_u32()),
                path: decomp.get_cstring()
            });
        }
        StorageInfo {
            data_hash,
            blocks,
            nodes
        }
    }
}
